stdlib/public/core/StringComparison.swift

```diff
@@ -0,0 +1,1164 @@
+    start: UnsafeMutableRawPointer(ptr).assumingMemoryBound(to: U.self),
+    count: count
+  )
+}
+@inline(__always)
+private func _unsafeBufferPointerCast<T, U>(
+  _ ptr: UnsafePointer<T>,
+  _ count: Int,
+  to: U.Type = U.self
+) -> UnsafeBufferPointer<U> {
+  return UnsafeBufferPointer(
+    start: UnsafeRawPointer(ptr).assumingMemoryBound(to: U.self),
+    count: count
+  )
+}
+
+internal let _leadingSurrogateBias: UInt16 = 0xd800
+internal let _trailingSurrogateBias: UInt16 = 0xdc00
+internal let _surrogateMask: UInt16 = 0xfc00
+
+@inline(__always)
+internal func _isSurrogate(_ cu: UInt16) -> Bool {
+  return _isLeadingSurrogate(cu) || _isTrailingSurrogate(cu)
+}
+
+@inline(__always)
+internal func _isLeadingSurrogate(_ cu: UInt16) -> Bool {
+  // NOTE: Specifically match against the trailing surrogate mask, as it matches
+  // more cases.
+  return cu & _surrogateMask == _leadingSurrogateBias
+}
+
+@inline(__always)
+internal func _isTrailingSurrogate(_ cu: UInt16) -> Bool {
+  return cu & _surrogateMask == _trailingSurrogateBias
+}
+@inline(__always)
+internal func _decodeSurrogatePair(
+  leading high: UInt16, trailing low: UInt16
+) -> UInt32 {
+  _sanityCheck(_isLeadingSurrogate(high) && _isTrailingSurrogate(low))
+  let hi10: UInt32 = UInt32(high) &- UInt32(_leadingSurrogateBias)
+  _sanityCheck(hi10 < 1<<10, "I said high 10. Not high, like, 20 or something")
+  let lo10: UInt32 = UInt32(low) &- UInt32(_trailingSurrogateBias)
+  _sanityCheck(lo10 < 1<<10, "I said low 10. Not low, like, 20 or something")
+
+  return ((hi10 &<< 10) | lo10) &+ 0x1_00_00
+}
+
+internal func _hasNormalizationBoundary(before cu: UInt16) -> Bool {
+  guard !_isSurrogate(cu) else { return false }
+  return UnicodeScalar(_unchecked: UInt32(cu))._hasNormalizationBoundaryBefore
+}
+
+//
+// Pointer casting helpers
+//
+internal func _castOutputBuffer(
+  _ ptr: UnsafeMutablePointer<_Normalization._SegmentOutputBuffer>,
+  endingAt endIdx: Int = _Normalization._SegmentOutputBuffer.capacity
+) -> UnsafeMutableBufferPointer<UInt16> {
+  let bufPtr: UnsafeMutableBufferPointer<UInt16> =
+    _unsafeMutableBufferPointerCast(
+      ptr, _Normalization._SegmentOutputBuffer.capacity)
+  return UnsafeMutableBufferPointer<UInt16>(rebasing: bufPtr[..<endIdx])
+}
+internal func _castOutputBuffer(
+  _ ptr: UnsafePointer<_Normalization._SegmentOutputBuffer>,
+  endingAt endIdx: Int = _Normalization._SegmentOutputBuffer.capacity
+) -> UnsafeBufferPointer<UInt16> {
+  let bufPtr: UnsafeBufferPointer<UInt16> =
+    _unsafeBufferPointerCast(
+      ptr, _Normalization._SegmentOutputBuffer.capacity)
+  return UnsafeBufferPointer<UInt16>(rebasing: bufPtr[..<endIdx])
+}
+
+extension _FixedArray16 where T == UInt16 {
+  mutating func fill(from other: _UnmanagedString<T>) {
+    _sanityCheck(other.count < _FixedArray16<T>.capacity,
+      "out of bounds fill")
+    for i in 0..<other.count {
+      self[i] = other[i]
+    }
+  }
+}
+
+public enum _Ordering: Int, Equatable {
```

Comment:
Make as much as you can `@_versioned internal` instead of public. Public may be need for testing